files/class-vip-filesystem.php

```diff
@@ -154,33 +161,45 @@ public function filter_upload_dir( $params ) {
 	}
 
 	/**
-	 * Check filetype support against VIP Filesystem API
-	 *
-	 * Leverages VIP Filesystem API, which will return a 406 or other non-200 code if the filetype is unsupported
-	 *
-	 * @since   1.0.0
-	 * @access  public
+	 * Validate the file before we allow uploading it.
 	 *
-	 * @param   array   $filetype_data
-	 * @param   string  $file
-	 * @param   string  $filename
-	 * @param   array   $mimes
+	 * @param  string[]  An array of data for a single file.
+	 */
+	public function filter_validate_file( $file ) {
+		$file_name = $file['name'];
+		$upload_path = trailingslashit( $this->get_upload_path() );
+		$file_path = $upload_path . $file_name;
+
+		// TODO: run through unique filename?
+
+		$check_type = $this->validate_file_type( $file_path );
+		if ( is_wp_error( $check_type ) ) {
+			$file['error'] = $check_type->get_error_message();
+
+			return $file;
+		}
+
+		$check_length = $this->validate_file_length( $file_path );
+		if ( is_wp_error( $check_length ) ) {
+			$file['error'] = $check_length->get_error_message();
+
+			return $file;
+		}
+
+		return $file;
+	}
+
+	/**
+	 * Check if file path in within the allowed length.
 	 *
-	 * @return  array
+	 * @param  string  Path starting with /wp-content/uploads
 	 */
-	public function filter_filetype_check( $filetype_data, $file, $filename, $mimes ) {
-		$filename = sanitize_file_name( $filename );
-
-		// Setting `ext` and `type` to empty will fail the upload because Go doesn't allow unfiltered uploads
-		// See `_wp_handle_upload()`
-		if ( ! $this->check_filetype_with_backend( $filename ) ) {
-			$filetype_data['ext']             = '';
-			$filetype_data['type']            = '';
-			// Never set this true, which leaves filename changing to dedicated methods in this class
-			$filetype_data['proper_filename'] = false;
+	protected function validate_file_length( $file_path ) {
```

Comment:
Minor thought: the name of this function may be a bit misleading because we are not checking the length of the file, but the length of its name instead.